django_emarsys/event.py

```diff
@@ -0,0 +1,263 @@
+        return num_new_events, num_updated_ids, num_deleted_ids, \
+            unknown_event_names, unsynced_event_names
+
+    for local_event in Event.objects.all():
+        emarsys_event_id = emarsys_events.get(local_event.name)
+        if emarsys_event_id:
+            if local_event.emarsys_id != emarsys_event_id:
+                num_updated_ids += 1
+
+            local_event.emarsys_id = emarsys_event_id
+            del emarsys_events[local_event.name]
+            local_event.save()
+        else:
+            local_event.delete()
+            num_deleted_ids += 1
+
+    for emarsys_event_name, emarsys_event_id in emarsys_events.items():
+        Event.objects.create(emarsys_id=emarsys_event_id,
+                             name=emarsys_event_name)
+        num_new_events += 1
+
+    all_synced_event_names = set(Event.objects.all()
+                                 .values_list('name', flat=True))
+    known_event_names = set(settings.EMARSYS_EVENTS.keys())
+    unknown_event_names = list(all_synced_event_names - known_event_names)
+    unsynced_event_names = list(known_event_names - all_synced_event_names)
+
+    if unknown_event_names:
+        log.warning("these Emarsys event names are not "
+                    "configured in settings.EMARSYS_EVENTS: {}"
+                    .format(', '.join(
+                        '"{}"'.format(x.replace('"', '\\"'))
+                            for x in unknown_event_names)))
+
+    if unsynced_event_names:
+        log.warning("these event names in settings.EMARSYS_EVENTS are not "
+                    "known by Emarsys: {}"
+                    .format(', '.join(
+                        '"{}"'.format(x.replace('"', '\\"'))
+                            for x in unsynced_event_names)))
+
+    return num_new_events, num_updated_ids, num_deleted_ids, \
+        unknown_event_names, unsynced_event_names
+
+
+def trigger_automatic_event(event_name, recipient_email, data=None,
+                            create_user_if_needed=True,
+                            contact_data_provider=None):
+    return _trigger_event(event_name=event_name,
+                          recipient_email=recipient_email,
+                          data=data,
+                          create_user_if_needed=create_user_if_needed,
+                          contact_data_provider=contact_data_provider,
+                          source=EventInstance.SOURCE_AUTOMATIC)
+
+
+def trigger_manual_event(event_name, recipient_email, data=None,
+                         create_user_if_needed=True,
+                         contact_data_provider=None):
+    return _trigger_event(event_name=event_name,
+                          recipient_email=recipient_email,
+                          data=data,
+                          create_user_if_needed=create_user_if_needed,
+                          contact_data_provider=contact_data_provider,
+                          source=EventInstance.SOURCE_MANUAL)
```

Comment:
Lately I feel strongly about taking everything away from the models :) that isn't directly related to their task, which is persistence. The models are the persistence layer. The persistence layer doesn't determine which possible sources there are, the business concept we're implementing is. For that reason I don't see "a way the model could change involving the source attribute" (quoted for sponginess ;)) that isn't triggered by a change in the business concept. Models entmachten! Views entmachten! Independent business logic for president!

So I'm for 3., and for pragmatic reasons I don't see why the persistence layer shouldn't use the constants from the business concept directly.

I don't insist on this change, but I hope that my arguments are convincing :)